Fix service crash when non-arsenal errors is sent to delete route

williamlardier opened the pull request:
Fixes a crash when the error we receive from the rout is not an Arsenal error (which is a possible and valid case).

Seragonia committed:
Fix crash when error is not an arsenal error in route

- Error may be returned by modules not using Arsenal
- In this case, any non-arsenal error reaching this code would
  cause a crash of the service.

Issue: ARSN-441

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-441 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.140`

* `8.2.0`


Please check the `Fix Version/s` of ARSN-441, or the target
branch of this pull request.

Seragonia committed:
Fix logger usage

Issue: ARSN-216, ARSN-441

Seragonia committed:
Extend the route types to support non-arsenal errors

Arsenal routes today are used in cloudserver, that calls these
callback functions using, most of the time, Arsenal errors. But
sometimes, especially when using modules, such as hdclient or
others, we might not return an Arsenal error. In this case, we
still try to access undefined properties, eventually leading to
strange responses to the client.

The changes are introducing support for traditional Error objects,
so that we can safely pass any error from a module used in an API
call.

Seragonia committed:
Implement unit tests for all routes

Issue: ARSN-441

Seragonia committed:
Test routesUtils

Issue: ARSN-441

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

welansari commented on lib/s3routes/routesUtils.ts:
```diff
@@ -717,30 +709,37 @@ export function streamUserErrorPage(
 ) {
+    let error;
+    if (err instanceof ArsenalError) {
+        error = err;
+    } else if (!(err instanceof ArsenalError)) {
+        error = errors.InternalError.customizeDescription(err.message);
```
```suggestion
        error = err;
    } else {
        error = errors.InternalError.customizeDescription(err.message);
```

welansari commented on lib/s3routes/routesUtils.ts:
```diff
@@ -106,34 +104,36 @@ const XMLResponseBackend = {
         setCommonResponseHeaders(corsHeaders, response, log);
+        let error;
+        if (errCode instanceof ArsenalError) {
+            error = errCode;
+        } else if (errCode instanceof Error) {
+            error = errors.InternalError.customizeDescription(errCode.message);
```
Is there a reason on why this check is inconsistent between the different functions of this file ?

williamlardier replied:
I'm not sure I understand fully the question, as we also have this logic in the other `errorResponse` for the json mode, but we work with `err.code` in the functions, something that we do not have if the error is not an Arsenal (or S3 standard) error. In this case, we want to recreate an internal error, so we have a logical error code, and we do not alter the implementation, while allowing any kind of error. Because we cannot guarantee that the error we receive will always be an Arsenal error (the implementation is modular and we use projects like hdclient that do not use Arsenal error, yet hdclient errors can reach this function).

welansari commented on lib/s3routes/routes/routeDELETE.ts:
```diff
@@ -54,11 +54,11 @@ export default function routeDELETE(
+                 * Since AWS expects a 204 regardless of the existence of
+                 * the object, the errors NoSuchKey and NoSuchVersion should not
+                 * be sent back as a response.
+                 */
+                if (err instanceof ArsenalError && !err.is?.NoSuchKey && !err.is?.NoSuchVersion) {
                     return routesUtils.responseNoBody(err, corsHeaders,
```
Should the non arsenal errors also be sent as a response ?

williamlardier replied:
No, because they would likely translate code-related error, leaking information about the internals of the application. Better returning standard errors in this case. But the routes should still allow these non-arsenal errors to be received, before sending the final response

francoisferrand commented on lib/s3routes/routes/routeDELETE.ts:
```diff
@@ -54,11 +54,11 @@ export default function routeDELETE(
+                 * the object, the errors NoSuchKey and NoSuchVersion should not
+                 * be sent back as a response.
+                 */
+                if (err instanceof ArsenalError && !err.is?.NoSuchKey && !err.is?.NoSuchVersion) {
```
* why would it not be an `ArsenalError` ? (i.e. are we not just "masking" an underlying issue: which should be fixed as well)
* if we receive an error, but not an ArsenalError : we change the behavior here... We probably don't want to include too much information indeed, but we should probably not send 204 either and report some sort of error (maybe a generic 500?)

williamlardier replied:
> why would it not be an ArsenalError ? (i.e. are we not just "masking" an underlying issue: which should be fixed as well)

Because we return errors from the underlying modules, such as hdclient, in our APIs. These errors are not Arsenal errors. It's possible for example in the "DeleteObject" API if hdclient returns an error. It does not happen most of the time, but it can happen, and we should not make assumption in Arsenal for the error type: this was changed when typescript was implemented in arsenal, but our S3 implementation was not updated to always return ArsenalErrors. And it actually not only return ArsenalErrors.

> if we receive an error, but not an ArsenalError : we change the behavior here... We probably don't want to include too much information indeed, but we should probably not send 204 either and report some sort of error (maybe a generic 500?)

I am not sure to understand what you mean by "we change the behavior here": its the actual bug we are trying to fix. Reaching this line with a non-Arsenal error makes the process crash, because `is` is not a valid field from a normal error object.

Example:
```
 stderr F 2024-07-12T02:51:37.386Z: Uncaught exception: 
 stderr F TypeError: Cannot read properties of undefined (reading 'NoSuchKey')
 stderr F     at /usr/src/app/node_modules/arsenal/build/lib/s3routes/routes/routeDELETE.js:82:32
 stderr F     at /usr/src/app/lib/api/objectDelete.js:251:20
 stderr F     at /usr/src/app/node_modules/async/dist/async.js:421:16
 stderr F     at next (/usr/src/app/node_modules/async/dist/async.js:5302:29)
 stderr F     at /usr/src/app/node_modules/async/dist/async.js:906:16
 stderr F     at /usr/src/app/lib/api/objectDelete.js:211:41
 stderr F     at /usr/src/app/lib/services.js:372:36
 stderr F     at /usr/src/app/node_modules/hdclient/lib/hdcontroller.js:358:17
 stderr F     at /usr/src/app/node_modules/async/dist/async.js:421:16
```

francoisferrand replied:
> I am not sure to understand what you mean by "we change the behavior here": it's the actual bug we are trying to fix

the previous behavior was to response w/o body (i.e. take the `if` branch) on any error except NoSuchKey / NoSuchVersion.

with the change, non-arsenal errors will not take this code path anymore : which is changing the behavior, not just fixing the crash...

the following would fix the crash without changing the behavior:

```suggestion
                if (err && !err.is?.NoSuchKey && !err.is?.NoSuchVersion) {
```

francoisferrand replied:
AFAIU, we can now have non-arsenal errors from Vault, right?
and we may want to have 204 in such case?

williamlardier replied:
We always call `routesUtils.responseNoBody` in both cases, the only difference being the error code returned, but that does change your suggestion indeed.
However we cannot do that, if we do not test that the error is an arsenal error we get TS errors, so I adapted it a bit.

-> Updated, I also removed the last bump commit as per our discussion

benzekrimaha commented on lib/s3routes/routes/routeGET.ts:
```diff
@@ -87,8 +87,7 @@ export default function routerGET(
-                    // TODO ARSN-216 Fix logger
-                    // @ts-ignore
+                    // @ts-expect-error werelogs types are not up to date
                     log.end().addDefaultFields({ contentLength });
```
shouldn't this be log.addDefaultFields({ contentLength }); like what's in here : https://github.com/scality/Arsenal/pull/2267/commits/449be66b448ec7574eab8edd8377c9f97037ef40#diff-8357453c6abb1c2118b5ddb61026fcce2f18a329f56eef87a68c0a83c3782e27R228

williamlardier replied:
It should actually be like in the get router, the end() returns "this" if no input is provided, but the types are not reporting it, so typescript thinks there is an error
https://github.com/scality/Arsenal/pull/2267/commits/8e830fe4b5f0dc34d243763a9bdf36f2cde40aa7

Seragonia committed:
End logger in Put route

Issue: ARSN-441

williamlardier commented:
/wait
/approve

Seragonia committed:
Bump werelogs

Issue: ARSN-441

Seragonia committed:
Go back to info logs after end() call following types fixes

Issue: ARSN-441

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-441 contains:


* `8.1.140`

* `8.2.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.141`

* `8.2.0`


Please check the `Fix Version/s` of ARSN-441, or the target
branch of this pull request.

ghost commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ARSN-441` into
`development/8.1`
* [w/8.2/bugfix/ARSN-441](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/bugfix/ARSN-441) will be merged into `development/8.2` 



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

Seragonia committed:
Ensure no behavior change with non arsenal errors processing

Issue: ARSN-441

bert-e commented:
# History mismatch



Merge commit #4f641951bf8210352d7b0cbc96e3872538f7a99c on the integration branch
`w/8.2/bugfix/ARSN-441` is merging a branch which is neither the current
branch `bugfix/ARSN-441` nor the development branch
`development/8.2`.

It is likely due to a rebase of the branch `bugfix/ARSN-441` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

ghost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ARSN-441` into
`development/8.1`
* [w/8.2/bugfix/ARSN-441](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/bugfix/ARSN-441) will be merged into `development/8.2` 



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

francoisferrand commented on lib/s3routes/routes/routeWebsite.ts:
```diff
@@ -30,7 +30,7 @@ export default function routerWebsite(
                 if (redirectInfo) {
                     if (err && redirectInfo.withError) {
-                        return routesUtils.redirectRequestOnError(err,
+                        return routesUtils.redirectRequestOnError(err as ArsenalError,
```
do you know why we need type coercion here?
`callApiMethod` already defines that the callback receives an `ArsenalError` error...

williamlardier replied:
It's because the error we can currently get from APIs is defined in

```ts

export type CallApiMethod = (
    methodName: string,
    request: http.IncomingMessage,
    response: http.ServerResponse,
    log: RequestLogger,
    callback: (err: ArsenalError | Error | null, ...data: any[]) => void,
) => void;
```

So it can be a non-arsenal error, this change was made on purpose so we benefit from ts checks everywhere and cover all paths that could lead to errors. The case of website is special, we already know it can only be arsenal errors. All of this will be removed once we have the "long term" solution in place

francoisferrand replied:
indeed, I missed the alternate types 🤦‍♂️

but it seems a bit dangerous to just "expect" these will be arsenal errors and blindly trust this is the case : if we want to rely on typescript to detect errors, best do it all the way and add an early check that it is indeed an ArsenalError (or just don't cast, and allow `redirectRequestOnError` to handle any `Error`)

williamlardier replied:
I updated the function so that it accepts non-arsenal errors - i kept the logic to dynamically recreate an arsenal error in this case, so we minimize the changes (we still need to return an Arsenal error today to the client).

francoisferrand replied:
👍, but then we should remove the type-coercion altogether

williamlardier replied:
yes... thanks

francoisferrand commented on lib/s3routes/routes/routeDELETE.ts:
```diff
@@ -54,11 +54,14 @@ export default function routeDELETE(
+                 */
+                if (err && (
+                    !(err instanceof ArsenalError) || 
+                    (err.is?.NoSuchKey === false && err.is?.NoSuchVersion === false)
```
nit: the `?` is not needed anymore in that case, since we are garanteed to have an `ArsenalError`.

(and best to use bool equivalence than strict equality to false, IMO...)

```suggestion
                    (!err.is.NoSuchKey && !err.is.NoSuchVersion)
```

williamlardier replied:
done

bert-e commented:
# History mismatch



Merge commit #6a61a1ef32940fb171ede13e0bce18dc581c3728 on the integration branch
`w/8.2/bugfix/ARSN-441` is merging a branch which is neither the current
branch `bugfix/ARSN-441` nor the development branch
`development/8.2`.

It is likely due to a rebase of the branch `bugfix/ARSN-441` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

ghost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ARSN-441` into
`development/8.1`
* [w/8.2/bugfix/ARSN-441](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/bugfix/ARSN-441) will be merged into `development/8.2` 



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

Seragonia committed:
Do not make assumption in routes for error instances

Issue: ARSN-441

Seragonia committed:
Remove type casting in website routes

Issue: ARSN-441

Seragonia committed:
Enforce using the arsenal error in route utils

Issue: ARSN-441

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue ARSN-441.

Goodbye williamlardier.




*The following options are set:* **approve, create_integration_branches**
